File: 2020/08/1.go
package main

import (
  "bufio"
  "fmt"
  "os"
  "strings"
  "strconv"
  "log"
)

type Instruction struct {
  name string
  arg int
}

func main() {
  fmt.Println(solution(prepareData(readInput("example_input"))))
  fmt.Println(solution(prepareData(readInput("input"))))
}

func solution(data []Instruction) (int) {
  executed := map[int]bool{}
  acc := 0
  idx := 0
  for idx < len(data) {
    instr := data[idx]
    if executed[idx] {
      return acc
    }
    executed[idx] = true
    if instr.name == "nop" {
      idx = idx + 1
      continue
    }
    if instr.name == "acc" {
      acc = acc + instr.arg
      idx = idx + 1
      continue
    }
    if instr.name == "jmp" {
      idx = idx + instr.arg
      continue
    }
    log.Fatal("Unknow instruction detected")
  }
  return acc
}

func prepareData(lines []string) ([]Instruction) {
  var prepared []Instruction
  for _, line := range lines {
    parts := strings.Split(strings.TrimSpace(line), " ")
    arg, err := strconv.Atoi(parts[1])
    if err != nil {
      log.Fatal(err)
    }
    prepared = append(prepared, Instruction{parts[0], arg})
  }
  return prepared
}

func readInput(path string) ([]string) {
  file, err := os.Open(path)
  if err != nil {
    log.Fatal(err)
  }
  defer file.Close()

  scanner := bufio.NewScanner(file)
  var lines []string
  for scanner.Scan() {
    lines = append(lines, scanner.Text())
  }

  if err := scanner.Err(); err != nil {
    log.Fatal(err)
  }

  return lines
}

File: 2020/09/1.py
def solution(data, preamble):
    sums = []
    for i, a in enumerate(data[:preamble]):
        for j, b in enumerate(data[:preamble]):
            if i == j:
                continue
            sums.append(a + b)

    for pointer, n in enumerate(data[preamble:]):
        if n not in sums:
            return n
        sums = sums[preamble:]
        sums = sums + [n + x for x in data[pointer:preamble + pointer]]

def read_input(file_path):
    with open(file_path, 'rt') as fi:
        return [prepare_input(line) for line in fi]

def prepare_input(line):
    return int(line.strip())

if __name__ == '__main__':
    print(solution(read_input('example_input'), 5))
    print(solution(read_input('input'), 25))
